native/src/core/module.rs

```diff
@@ -0,0 +1,572 @@
+                                FsNode::Symlink {
+                                    target: link.to_owned(),
+                                },
+                            );
+                        } else {
+                            mount_dummy("mirror", path.real, path.tmp, entry.is_dir())?;
+                        }
+                    }
+                }
+
+                // Finally, commit children
+                for (name, node) in children {
+                    let path = path.append(name);
+                    node.commit_tmpfs(path)?;
+                }
+            }
+            FsNode::File { src } => {
+                if path.real.exists() {
+                    clone_attr(path.real, src)?;
+                }
+                mount_dummy("mount", src, path.tmp, false)?;
+            }
+            FsNode::Symlink { target } => {
+                module_log!("mklink", path.tmp, target);
+                path.tmp.create_symlink_to(target)?;
+                if path.real.exists() {
+                    clone_attr(path.real, path.tmp)?;
+                }
+            }
+            FsNode::Whiteout => {
+                module_log!("delete", path.real, "null");
+            }
+        }
+        Ok(())
+    }
+}
+
+fn get_path_env() -> String {
+    std::env::var_os("PATH")
+        .and_then(|s| s.into_string().ok())
+        .unwrap_or_default()
+}
+
+fn inject_magisk_bins(system: &mut FsNode) {
+    fn inject(children: &mut FsNodeMap) {
+        let mut path = cstr::buf::default().join_path(get_magisk_tmp());
+
+        // Inject binaries
+
+        let len = path.len();
+        path.append_path("magisk");
+        children.insert(
+            "magisk".to_string(),
+            FsNode::File {
+                src: path.to_owned(),
+            },
+        );
+
+        path.truncate(len);
+        path.append_path("magiskpolicy");
+        children.insert(
+            "magiskpolicy".to_string(),
+            FsNode::File {
+                src: path.to_owned(),
+            },
+        );
+
+        // Inject applet symlinks
+
+        children.insert(
+            "su".to_string(),
+            FsNode::Symlink {
+                target: Utf8CString::from("./magisk"),
+            },
+        );
+        children.insert(
+            "resetprop".to_string(),
+            FsNode::Symlink {
+                target: Utf8CString::from("./magisk"),
+            },
+        );
+        children.insert(
+            "supolicy".to_string(),
+            FsNode::Symlink {
+                target: Utf8CString::from("./magiskpolicy"),
+            },
+        );
+    }
+
+    // First find whether /system/bin exists
+    let bin = system.children().and_then(|c| c.get_mut("bin"));
+    if let Some(FsNode::Directory { children }) = bin {
+        inject(children);
+        return;
+    }
+
+    // If /system/bin node does not exist, use the first suitable directory in PATH
```

Comment:
Maybe we can use `/system/apex/com.android.runtime/bin`, all custom ROMs are flattened apex.
https://source.android.com/docs/core/ota/apex#flattening-apex